# -*- coding: utf-8 -*-
from setuptools import setup

package_dir = \
{'': 'src'}

packages = \
['breakfast']

package_data = \
{'': ['*']}

install_requires = \
['click>=8.1.3,<9.0.0',
 'networkx>=2.8,<3.0',
 'numpy>=1.22.3,<2.0.0',
 'pandas>=1.4.2,<2.0.0',
 'scikit-learn>=1.0.2,<2.0.0',
 'scipy>=1.8.0,<2.0.0']

entry_points = \
{'console_scripts': ['breakfast = breakfast.console:main']}

setup_kwargs = {
    'name': 'breakfast',
    'version': '0.4.3',
    'description': 'breakfast: fast putative outbreak cluster and infection chain detection using SNPs',
    'long_description': '# breakfast - FAST outBREAK detection and sequence clustering\n\n[![Tests](https://github.com/rki-mf1/breakfast/workflows/Tests/badge.svg)](https://github.com/rki-mf1/breakfast/actions?workflow=Tests)\n\n`breakfast` is a simple and fast script developed for clustering SARS-CoV-2 genomes using precalculated sequence features (e.g. nucleotide substitutions) from [covSonar](https://github.com/rki-mf1/covsonar) or [Nextclade](https://clades.nextstrain.org/).\n\n**This project is under development and in experimental stage**\n\n<img src="/img/breakfast_logo_2.png" width="300">\n\n## Installation\n\n### Installation using conda/mamba\n\nBreakfast is available in [bioconda](http://bioconda.github.io/recipes/breakfast/README.html). You can install it using either the conda command, or if you\'ve installed [mamba](https://github.com/mamba-org/mamba) you can use that:\n\n```\n$ conda install breakfast\n# or\n$ mamba install breakfast\n```\n\n### Installation using pip\n\nConda is available from [PyPI](https://pypi.org/project/breakfast/) and can be installed using pip:\n\n```\n$ pip install breakfast\n```\n\n## Example Command Line Usage\n\n### Simple test run\n\n```\nbreakfast \\\n   --input-file breakfast/test/testfile.tsv  \\\n   --max-dist 1 \\\n   --outdir test-run/\n```\n\nYou will find your results in `test-run/cluster.tsv`, which should be identical to `breakfast/test/expected_clusters_dist1.tsv`\n\n### Using breakfast with input from covsonar\n\nbreakfast uses pre-calculated sequence features (= mutations) as input rather than raw sequences. These features can be calculated with several different programs, but the one we mainly use is [covSonar](https://github.com/rki-mf1/covsonar). It can be used to maintain a database of mutations for a large number of sequences, which can then be easily queried.\n\n```\nconda activate sonar\ncovsonar/sonar.py add -f genomes.fasta --db mydb --cpus 8\ncovsonar/sonar.py match --tsv --db mydb > genomic_profiles.tsv\n```\n\nClustering with a maximum SNP-distance of 1 and excluding clusters below a size of 5 sequences:\n\n```\nbreakfast \\\n   --input-file genomic_profiles.tsv \\\n   --max-dist 1 \\\n   --min-cluster-size 5 \\\n   --outdir covsonar-breakfast-results/\n```\n\n### Using breakfast with input from Nextclade\n\nAn alternative to covsonar that is commonly used is [Nextclade CLI](https://clades.nextstrain.org/).\n\n```\nconda install -c bioconda nextclade  # If nextclade isn\'t already installed\nnextclade dataset get --name \'sars-cov-2\' --output-dir \'data/sars-cov-2\'\nnextclade \\\n   --in-order \\\n   --input-fasta genomes.fasta \\\n   --input-dataset data/sars-cov-2 \\\n   --output-tsv output/nextclade.tsv \\\n   --output-tree output/nextclade.auspice.json \\\n   --output-dir output/ \\\n   --output-basename nextclade\n```\n\nAlternatively, you can also use [Nextclade Web](https://clades.nextstrain.org/) to process your fasta and export the genomic profile as "nextclade.tsv".\n\nClustering with a maximum SNP-distance of 1 and excluding clusters below a size of 5 sequences. Since the input tsv of Nextclade looks a little different from the covSonar tsv, you need to specify the additional parameters `--id-col`, `--clust-col` and `--sep2` for identifying the correct columns.\n\n```\nbreakfast \\\n   --input-file output/nextclade.tsv \\\n   --max-dist 1 \\\n   --min-cluster-size 5 \\\n   --id-col "seqName" \\\n   --clust-col "substitutions" \\\n   --sep2 "," \\\n   --outdir nextclade-breakfast-results/\n```\n\n## Sequence feature formats\n\nTypical input data to breakfast looks something like the following table (example from covsonar with unnecessary columns removed):\n\n```\naccession\tdna_profile\nexample1\tC241T T606C C913T C3037T del:11288:9 C13515T\nexample2\tC241T T606C del:1000:10\nexample3\tC241T T606C del:1001:20\n```\n\nbreakfast has parameters to allow the user to ignore deletions (`--skip-del`), insertions (`--skip-ins`) or mutations at the end of the sequences (which can sometimes be error-prone) when calculating the distance between sequences. To be able to do this, we need to know what kind of input is being provided (using the `--var-type` option), and then parse the mutations themselves. Since the format of how mutations are represented by upstream programs differs, we have implemented program-specific parsers for covsonar DNA and AA, as well as nextclade DNA and AA. Examples are shown below, in case you want to use some other program as input to breakfast you can see if the format matches one of the existing feature formats. As a fallback, you can use the "raw" format, which disables parsing and does not allow you to use breakfast\'s indel skipping or trimming features.\n\n### covsonar\n\n| mutation type | DNA (`covsonar_dna`) | AA (`covsonar_aa`) |\n| ------------ | ----- | ------- |\n| substitution | C241T | S:N501Y |\n| deletion | del:11288:9 | ORF1:del:12:7 |\n| insertion | C241TAT | N:A34AK |\n\n### Nextclade\n\n| mutation type | DNA (`nextclade_dna`) | AA (`nextclade_aa`) |\n| ------------ | ----- | ------- |\n| substitution | C241T | S:N501Y |\n| deletion | 11288-11297 or 22492 | S:V70- |\n| insertion | 273:CTTCGA | (not provided) |\n\n### Raw\n\nFeatures will not be parsed. Skipping inserts (`--skip-ins`) and/or deletions (`--skip-del`) and the trimming options are not supported with this feature type.\n\n## Parameter description\n\n| Parameter              | Type    \t| Required | Default \t| Description                                |\n|----------------------- |---------\t|----------|----------|------------------------------------------  |\n| --input-file           | String     \t|âœ…\t     | \'genomic_profiles.tsv.gz\'    \t| Path of the input file (in tsv format)     |\n| --max-dist              | Integer  \t|\t     | 1     \t| Two sequences will be grouped together, if their pairwise edit distance does not exceed this threshold |\n| --min-cluster-size  | Integer  \t|      | 2     \t| Minimum number of sequences a cluster needs to include to be defined in the result file      |\n| --id-col    | String \t|     | \'accession\'      \t| Name of the sequence identifier column of the input file          |\n| --clust-col              | String \t|     | \'dna_profile\'      | Name of the mutation profile column of the input file         |\n| --var-type              | String \t|     | \'covsonar_dna\'       | Mutuation format (e.g. for DNA mutations from covsonar, use covsonar_dna). Possible values: [covsonar_dna|covsonar_aa|nextclade_dna|nextclade_aa|raw] |\n| --sep              | String \t|     | \'\\t\'      | Input file separator       |\n| --sep2              | String \t|     | \' \'      | Secondary clustering column separator (between each mutation)        |\n| --outdir              | String \t|     | \'output/\'       | Path of output directory        |\n| --trim-start               | Integer \t|     | 264       | Bases to trim from the beginning         |\n| --trim-end               | Integer \t|     | 228       | Bases to trim from the end         |\n| --reference-length              | Integer \t|     | 29903      | Length of reference genome (defaults to NC_045512.2)        |\n| --skip-del               | Bool \t|     | TRUE       | Deletions will be skipped for calculating the pairwise distance of your input sequences.|\n| --skip-ins               | Bool \t|     | TRUE       | Insertions will be skipped for calculating the pairwise distance of your input sequences.         |\n| --input-cache           | Integer \t|     | None   | Path to import results from previous run |\n| --output-cache              | String \t|     | None       | Path to export results which can be used in the next run to decrease runtime.  |\n| --jobs              | Integer \t|     | 1       | The number of jobs (=threads) to run simultaneously |\n| --help                   | N/A     \t|\t   | N/A     \t| Show this help message and exit            |\n| --version                | N/A     \t|\t   | N/A     \t| Show version and exit            |\n\n## Dependencies\n\n`breakfast` runs under Python 3.10 and later. We rely heavily on some excellent open source python libraries: networkx, pandas, numpy, scikit-learn, click, and scipy.\n',
    'author': 'Matthew Huska',
    'author_email': 'HuskaM@rki.de',
    'maintainer': None,
    'maintainer_email': None,
    'url': 'https://github.com/rki-mf1/breakfast',
    'package_dir': package_dir,
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.9,<3.11',
}


setup(**setup_kwargs)
